src/raft.rs

```diff
@@ -555,19 +563,45 @@ impl<T: Storage> Raft<T> {
         m.set_entries(RepeatedField::from_vec(ents));
         m.set_commit(self.raft_log.committed);
         if !m.get_entries().is_empty() {
-            match pr.state {
-                ProgressState::Replicate => {
-                    let last = m.get_entries().last().unwrap().get_index();
-                    pr.optimistic_update(last);
-                    pr.ins.add(last);
+            let last = m.get_entries().last().unwrap().get_index();
+            pr.update_state(last);
+        }
+    }
+
+    fn try_batching(
+        &mut self,
+        to: u64,
+        pr: &mut Progress,
+        ents: &mut Vec<Entry>,
+        term: u64,
+    ) -> bool {
+        // if MsgAppend for the reciver already exists, try_batching
+        // will append the entries to the existing MsgAppend
+        let mut is_batched = false;
+        let mut last_idx = 0;
+        let mut is_empty = true;
+        for msg in &mut self.msgs {
+            if msg.get_log_term() == term
+                && msg.get_msg_type() == MessageType::MsgAppend
+                && msg.get_to() == to
+            {
+                is_batched = true;
+                msg.set_index(pr.next_idx - 1);
+                msg.set_commit(self.raft_log.committed);
+                let mut batched_entries = msg.take_entries().into_vec();
+                batched_entries.append(ents);
```

Comment:
What if logs in origin msg and in ents are not continuous?

Reply:
will that case come? (sorry if I've asked stupid question :( )

Reply:
Yes, follower can reject the leader's `MsgAppend` to make it reset `next_index`.

Reply:
If it that case should I discard the previous ents?

Reply:
Yes. For simplicity, you can also just give up batching in that case.